packages/fxa-settings/src/lib/error-utils.ts

```diff
@@ -0,0 +1,175 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+import { GraphQLError } from 'graphql';
+import {
+  AuthUiError,
+  AuthUiErrorNos,
+  AuthUiErrors,
+} from './auth-errors/auth-errors';
+import VerificationMethods from '../constants/verification-methods';
+import VerificationReasons from '../constants/verification-reasons';
+import * as Sentry from '@sentry/browser';
+import { FtlMsgResolver } from 'fxa-react/lib/utils';
+import { OAUTH_ERRORS } from './oauth';
+
+// TODO: Consolidate with AuthUiError type
+export interface GenericError {
+  errno: number;
+  message: string;
+  /** Not typically used in the UI, prefer 'errno' instead */
+  code?: number;
+  retryAfter?: number;
+  retryAfterLocalized?: string;
+}
+
+export type BeginSigninError = GenericError & {
+  verificationReason?: VerificationReasons;
+  verificationMethod?: VerificationMethods;
+  /** Important for 'Incorrect email case' edgecase */
+  email?: string;
+};
+
+export type HandledError = GenericError | BeginSigninError;
+
+const handleAuthUiError = (error: {
+  errno: number;
+  message: string;
+}): { error: HandledError } => {
+  const { errno } = error as HandledError;
+  if (errno && AuthUiErrorNos[errno]) {
+    return { error };
+  }
+  return { error: AuthUiErrors.UNEXPECTED_ERROR as HandledError };
+};
+
+export const getHandledError = (error: {
+  errno: number;
+  message: string;
+  graphQLErrors?: GraphQLError[];
+}) => {
+  const graphQLError: GraphQLError | undefined = error.graphQLErrors?.[0];
+  if (graphQLError) {
+    return handleGQLError(graphQLError);
+  }
+  return handleAuthUiError(error);
+};
+
+const handleGQLError = (graphQLError: GraphQLError) => {
+  const errno = graphQLError.extensions.errno as number;
+
+  if (errno && AuthUiErrorNos[errno]) {
+    const uiError = {
+      message: AuthUiErrorNos[errno].message,
+      errno,
+      email:
+        errno === AuthUiErrors.INCORRECT_EMAIL_CASE.errno
+          ? graphQLError.extensions.email
+          : undefined,
+      verificationMethod:
+        (graphQLError.extensions.verificationMethod as VerificationMethods) ||
+        undefined,
+      verificationReason:
+        (graphQLError.extensions.verificationReason as VerificationReasons) ||
+        undefined,
+      retryAfter: (graphQLError.extensions.retryAfter as number) || undefined,
+      retryAfterLocalized:
+        (graphQLError?.extensions.retryAfterLocalized as string) || undefined,
```

Comment:
Q: Does this need a `?.` after `graphQLError`? We don't seem to have it on L60, L68, L71, L74, or L76.

Reply:
It does not, thanks for the catch.